backend/amd64/cfg_stack_operands.ml

```diff
@@ -0,0 +1,218 @@
+  = fun map instr ->
+  if debug then check_lengths instr ~of_arg:0 ~of_res:1;
+  begin match is_spilled instr.res.(0) with
+  | false ->
+    All_spilled_registers_rewritten
+  | true ->
+    use_stack_operand map instr.res 0;
+    All_spilled_registers_rewritten
+  end
+
+let may_use_stack_operand_for_result
+  : type a . num_args:int -> Reg.t Reg.Tbl.t -> a Cfg.instruction -> stack_operands_rewrite
+  = fun ~num_args map instr ->
+  if debug then begin
+    check_lengths instr ~of_arg:num_args ~of_res:1;
+  end;
+  begin match is_spilled instr.res.(0) with
+  | false -> ()
+  | true ->
+    if Reg.same instr.arg.(0) instr.res.(0) then begin
+      use_stack_operand map instr.arg 0;
+    end;
+    use_stack_operand map instr.res 0;
+  end;
+  May_still_have_spilled_registers
+
+type result =
+  | No_result
+  | Result_can_be_on_stack
+  | Result_cannot_be_on_stack
+
+let binary_operation
+  : type a . Reg.t Reg.Tbl.t -> a Cfg.instruction -> result -> stack_operands_rewrite
+  = fun map instr result ->
+  if debug then begin
+    match result with
+    | No_result ->
+      check_lengths instr ~of_arg:2 ~of_res:0
+    | Result_can_be_on_stack ->
+      check_lengths instr ~of_arg:2 ~of_res:1;
+      check_same "res(0)" instr.res.(0) "arg(0)" instr.arg.(0)
+    | Result_cannot_be_on_stack ->
+      check_lengths instr ~of_arg:2 ~of_res:1
+  end;
+  begin match is_spilled instr.arg.(0), is_spilled instr.arg.(1) with
+  | false, false ->
+    begin match result with
+    | No_result | Result_can_be_on_stack ->
+      All_spilled_registers_rewritten
+    | Result_cannot_be_on_stack ->
+      May_still_have_spilled_registers
+    end
+  | false, true ->
+    use_stack_operand map instr.arg 1;
+    begin match result with
+    | No_result ->
+      All_spilled_registers_rewritten
+    | Result_can_be_on_stack | Result_cannot_be_on_stack ->
+      May_still_have_spilled_registers
+    end
+  | true, false ->
+    begin match result with
+    | No_result ->
+      use_stack_operand map instr.arg 0;
+      All_spilled_registers_rewritten
+    | Result_can_be_on_stack ->
+      use_stack_operand map instr.res 0;
+      use_stack_operand map instr.arg 0;
+      All_spilled_registers_rewritten
+    | Result_cannot_be_on_stack ->
+      use_stack_operand map instr.arg 0;
+      May_still_have_spilled_registers
+    end;
+  | true, true ->
+    if Reg.same instr.arg.(0) instr.arg.(1) then
+      May_still_have_spilled_registers
+    else begin
+      match result with
+      | No_result ->
+        (* CR xclerc for xclerc: try and find for a criterion to choose between
+           the two. *)
+        use_stack_operand map instr.arg 0;
+        May_still_have_spilled_registers
+      | Result_can_be_on_stack ->
+        use_stack_operand map instr.arg 0;
+        use_stack_operand map instr.res 0;
+        May_still_have_spilled_registers
+      | Result_cannot_be_on_stack ->
+        use_stack_operand map instr.arg 1;
```

Comment:
I think this should keep arg.(0) on the stack instead, to match the existing amd64/reload  behavior but I'm not sure which instructions fall under this case, it's a bit hard to see form this code (compared to amd64/{selection,reload}.ml where the constraints for each instruction were all in one place, but I don't know how to improve the code here). Perhaps we should assert these architectural constraints, maybe in `basic`, and in particular the sharing constraints that are very fragile and easy to break when rewriting instructions?

Reply:
I will need to have another look; here arg.(0) and
res.(0) are the same, so I am not sure it is right to
change one but not the other.

Reply:
There is no check that arg.(0) and res.(0) are the same in the case of  `Result_cannot_be_on_stack`, is there?

Reply:
It looks like the only instructions that reach this case are Iintop (Icomp _), and in amd64/reload.ml  this would result in arg.(0) passed on the stack, arg.(1) in reg, unlike here.

Reply:
> There is no check that arg.(0) and res.(0) are the same in the case of `Result_cannot_be_on_stack`, is there?

No, but I don't think that property holds.

Reply:
yes, I agree. I thought that you said in the [comment above](https://github.com/ocaml-flambda/flambda-backend/pull/771#discussion_r967106530) that they are the same here.

Reply:
> It looks like the only instructions that reach this case are Iintop (Icomp _), and in amd64/reload.ml this would result in arg.(0) passed on the stack, arg.(1) in reg, unlike here.

Can you elaborate on what change you would
like to see, in terms of code?

Reply:
```suggestion
        use_stack_operand map instr.arg 0;
```